cmd/kubeapps-apis/plugins/kapp_controller/packages/v1alpha1/server_ctrl_packages.go

```diff
@@ -11,3 +11,119 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 package main
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	corev1 "github.com/kubeapps/kubeapps/cmd/kubeapps-apis/gen/core/packages/v1alpha1"
+	datapackagingv1alpha1 "github.com/vmware-tanzu/carvel-kapp-controller/pkg/apiserver/apis/datapackaging/v1alpha1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	log "k8s.io/klog/v2"
+)
+
+// GetAvailablePackageSummaries returns the available packages managed by the 'kapp_controller' plugin
+func (s *Server) GetAvailablePackageSummaries(ctx context.Context, request *corev1.GetAvailablePackageSummariesRequest) (*corev1.GetAvailablePackageSummariesResponse, error) {
+	log.Infof("+kapp-controller GetAvailablePackageSummaries")
+
+	// Retrieve the proper parameters from the request
+	namespace := request.GetContext().GetNamespace()
+	cluster := request.GetContext().GetCluster()
+	pageSize := request.GetPaginationOptions().GetPageSize()
+	pageOffset, err := pageOffsetFromPageToken(request.GetPaginationOptions().GetPageToken())
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "unable to intepret page token %q: %v", request.GetPaginationOptions().GetPageToken(), err)
+	}
+	// Assume the default cluster if none is specified
+	if cluster == "" {
+		cluster = s.globalPackagingCluster
+	}
+	// fetch all the package metadatas
+	pkgMetadatas, err := s.getPkgMetadatas(ctx, cluster, namespace)
```

Comment:
Given that the `PackageMetadata` contains the in-lined SVG icon, this call could itself be causing performance/mem issues for a repo like Bitnami? Worth checking and if so, only request the page size number of results.

Reply:
I've quickly analyzed the elapsed times of each part: 

- Get all package metadatas: 0.2 seconds, it's ok
- getPkgsWithFieldSelector: **2 seconds avg**
  - The "problem" is in the `resource.List(ctx, listOptions)` code, which is taking from 0.8 to 2.4 seconds in retrieving all the packages whose FieldSelector is foo. Some examples:
    - `resource.List function took 2.245123407s for (3) items for spec.refName=contour.community.tanzu.vmware.com`
    -  `resource.List function took 1.550673668s for (3) items for spec.refName=pkg.test.carvel.dev`
- getPkgVersionsMap function: microseconds, not relevant
- Build the  AvailablePackageSummary object: microseconds, not relevant


Adding a TODO linked to this comment.

Reply:
Excellent - thanks for the info.